tests/EdgeToEdge/Routes/ShowMembershipConfirmationRouteTest.php

```diff
@@ -36,44 +39,124 @@ private function newStoredMembershipApplication( FunFunFactory $factory ): Appli
 					'accessToken' => self::CORRECT_ACCESS_TOKEN
 				]
 			);
 
 			$this->assertSame(
 				'some value',
-				$client->getCookieJar()->get( ShowDonationConfirmationHandler::SUBMISSION_COOKIE_NAME )->getValue()
+				$client->getCookieJar()->get( ApplyForMembershipHandler::SUBMISSION_COOKIE_NAME )->getValue()
 			);
 		} );
 	}
 
 	public function testCallWithWrongAccessToken_deniedPageIsShown(): void {
 		$this->createEnvironment( [], function ( Client $client, FunFunFactory $factory ): void {
-			$donation = $this->newStoredMembershipApplication( $factory );
+			$membershipApplication = $this->newStoredMembershipApplication( $factory );
 
 			$client->request(
 				Request::METHOD_GET,
-				self::PATH, [
-					'id' => $donation->getId(),
+				self::PATH,
+				[
+					'id' => $membershipApplication->getId(),
 					'accessToken' => self::WRONG_ACCESS_TOKEN
 				]
 			);
 
-			$this->assertContains( 'access_denied_membership_confirmation', $client->getResponse()->getContent() );
-			$this->assertSame( 403, $client->getResponse()->getStatusCode() );
+			$this->assertAccessIsDenied( 'access_denied_membership_confirmation', $client );
+		} );
+	}
+
+	public function testCallOnAnonymizedRecord_deniedPageIsShown(): void {
+		$this->createEnvironment( [], function ( Client $client, FunFunFactory $factory ): void {
+			$factory->setMembershipTokenGenerator( new FixedMembershipTokenGenerator(
+				self::CORRECT_ACCESS_TOKEN
+			) );
+
+			$doctrineApplication = ValidMembershipApplication::newAnonymizedDoctrineEntity();
+
+			$factory->getEntityManager()->persist( $doctrineApplication );
+			$factory->getEntityManager()->flush();
+
+			$client->request(
+				Request::METHOD_GET,
+				self::PATH,
+				[
+					'id' => $doctrineApplication->getId(),
+					'accessToken' => self::CORRECT_ACCESS_TOKEN
+				]
+			);
+
+			$this->assertAccessIsDenied( 'access_denied_membership_confirmation_anonymized', $client );
 		} );
 	}
 
+	public function testCallOnUnknownApplicationId_deniedPageIsShown(): void {
+		$this->createEnvironment( [], function ( Client $client, FunFunFactory $factory ): void {
+			$client->request(
+				Request::METHOD_GET,
+				self::PATH,
+				[
+					'id' => self::UNKNOWN_APPLICATION_ID,
+					'accessToken' => self::CORRECT_ACCESS_TOKEN
+				]
+			);
+
+			$this->assertAccessIsDenied( 'access_denied_membership_confirmation', $client );
+		} );
+	}
+
+	public function testOnDatabaseError_errorPageIsShown(): void {
+		$membershipApplication = ValidMembershipApplication::newDomainEntity();
+
+		$client = $this->createClient( [], function ( FunFunFactory $factory ) use ( $membershipApplication ): void {
+
+			$factory->setMembershipTokenGenerator( new FixedMembershipTokenGenerator(
+				self::CORRECT_ACCESS_TOKEN
+			) );
+
+			$applicationRepository = $this->getMockBuilder( DoctrineApplicationRepository::class )
+				->setConstructorArgs( [$factory->getEntityManager()] )
+				->setMethods( ['getDoctrineApplicationById'] )
+				->getMock();
```

Comment:
What does this do? Does this mean only this method is mocked, so the other real production code methods are still there? If so this seems sane enough.

Beware that there might be a test double that does something similar. Perhaps a ThrowingApplicationRepository or an InMemoryApplicationRepository that has a throwOnRead method or some such. The approach used here has the drawback that you refer to a method name with a string (breaks refactoring and navigation such as find usages). It also makes the test bind to ORMException which on first glance looks to be an implementation detail of a specific implementation of the service. ... Yeah having thought about it a bit this coupling is definitely unfortunate.